include/triggeralgs/AbstractFactory.hxx

```diff
@@ -0,0 +1,61 @@
+/* @file: AbstractFactory.hxx
+ *
+ * This is part of the DUNE DAQ Application Framework, copyright 2023.
+ * Licensing/copyright details are in the COPYING file that you should have
+ * received with this code.
+ */
+
+#ifndef TRIGGERALGS_ABSTRACT_FACTORY_HXX_
+#define TRIGGERALGS_ABSTRACT_FACTORY_HXX_
+
+namespace triggeralgs {
+
+template <typename T>
+std::shared_ptr<AbstractFactory<T>> AbstractFactory<T>::s_single_factory = nullptr;
+
+template <typename T>
+typename AbstractFactory<T>::creation_map& AbstractFactory<T>::get_makers(){
+  static creation_map s_makers;
+  return s_makers;
+}
+
+template <typename T>
+void AbstractFactory<T>::register_creator(const std::string alg_name, maker_creator creator)
+{
+  creation_map& makers = get_makers();
+  auto it = makers.find(alg_name);
+
+  if (it == makers.end()) {
+    makers[alg_name] = creator;
+    return;
+  }
```

Comment:
I would add some error/debugging message in `else{}`, because if it did find the algorithm, then someone implemented algo with the same class name as another, or something like that... which they shouldn't, so this could crash to force them not to do it.